src/Slave.hpp

```diff
@@ -0,0 +1,148 @@
+            return m_can_open.download(T::OBJECT_ID, T::OBJECT_SUB_ID,
+                                       get<T>());
+        }
+
+        /** Query the download (= write to slave) of an object
+         *
+         * @arg offsetId an offset that should be applied to the object ID
+         * @arg offsetSubId an offset that should be applied to the object sub ID
+         */
+        template<typename T>
+        canbus::Message queryDownload(typename T::OBJECT_TYPE value,
+                                      int offsetId = 0, int offsetSubId = 0) const {
+            return m_can_open.download(T::OBJECT_ID + offsetId,
+                                       T::OBJECT_SUB_ID + offsetSubId,
+                                       value);
+        }
+
+        /** Get an object from the object database
+         */
+        template<typename T>
+        typename T::OBJECT_TYPE get(int offsetId = 0, int offsetSubId = 0) const {
+            return m_can_open.get<typename T::OBJECT_TYPE>(
+                T::OBJECT_ID + offsetId, T::OBJECT_SUB_ID + offsetSubId
+            );
+        }
+
+        /** Check whether the given object has been initialized in the object database
+         *
+         * @arg offsetId an offset that should be applied to the object ID
+         * @arg offsetSubId an offset that should be applied to the object sub ID
+         */
+        template<typename T>
+        bool has(int offsetId = 0, int offsetSubId = 0) const {
+            return m_can_open.has(T::OBJECT_ID + offsetId,
+                                  T::OBJECT_SUB_ID + offsetSubId);
+        }
+
+        /** Timestamp of the last written value for the given object (might be zero)
+         *
+         * @arg offsetId an offset that should be applied to the object ID
+         * @arg offsetSubId an offset that should be applied to the object sub ID
+         */
+        template<typename T>
+        base::Time timestamp(int offsetId = 0, int offsetSubId = 0) const {
+            return m_can_open.timestamp(T::OBJECT_ID + offsetId,
+                                        T::OBJECT_SUB_ID + offsetSubId);
+        }
+
+        template<typename T>
+        void set(typename T::OBJECT_TYPE value,
+                 base::Time const& time = base::Time::now()) {
+            return m_can_open.set<typename T::OBJECT_TYPE>(
+                T::OBJECT_ID, T::OBJECT_SUB_ID,
+                value, time
+            );
+        }
+
+        template<typename T>
+        void set(typename T::OBJECT_TYPE value,
+                 int offsetId, int offsetSubId = 0,
+                 base::Time const& time = base::Time::now()) {
+            return m_can_open.set<typename T::OBJECT_TYPE>(
+                T::OBJECT_ID + offsetId, T::OBJECT_SUB_ID + offsetSubId,
+                value, time
+            );
+        }
+
+        virtual Update process(canbus::Message const& message);
+
+        StateMachine& getStateMachine();
```

Comment:
I could also put the member public.

The class is, in the end, a higher level access API over the StateMachine. This is needed so that we don't have to necessarily reflect all possible use-cases of StateMachine (i.e. "go to the metal" if needed)

The alternative would be to force the user to pass the state machine to the constructor. I could also do that if it makes you feel more comfortable.

Reply:
> The alternative would be to make the user to pass the state machine to the constructor. I could also do that if it makes you feel more comfortable.

I think that makes more clear that the `StateMachine` is not owned by or private to the object. What do you think?

Reply:
I find the construction (in case someone only wants a Slave) a little bit awkward, but I agree that it makes the object relationships more obvious. So, no opinion. I'll follow your lead.

Reply:
Updated